src/babel/index.js

```diff
@@ -700,6 +712,93 @@ function transformZacsAttributesOnNonZacsElement(t, platform, path) {
     .concat(addedAttrs)
 }
 
+function isPlainObjectProperty(t, node) {
+  return (
+    t.isObjectProperty(node) &&
+    t.isIdentifier(node.key) &&
+    !node.shorthand &&
+    !node.computed &&
+    !node.method
+  )
+}
+
+function validateStyleSheet(t, args, path) {
+  const [stylesheet] = args
+  if (!(args.length === 1 && t.isObjectExpression(stylesheet))) {
+    throw path.buildCodeFrameError('ZACS StyleSheet accepts a single Object argument')
+  }
+  const stylesets = stylesheet.properties
+  if (!stylesets.every(styleset => isPlainObjectProperty(t, styleset))) {
+    // TODO: We can probably allow `"name":`, no problem.
+    throw path.buildCodeFrameError(
+      'ZACS StyleSheet stylesets must be defined as `name: {}`. Other syntaxes, like `[name]:`, `"name": `, `...styles` are not allowed',
+    )
+  }
+
+  if (
+    !stylesets.every(
+      styleset =>
+        t.isObjectExpression(styleset.value) &&
+        styleset.value.properties.every(property => isPlainObjectProperty(t, property)),
+    )
+  ) {
+    throw path.buildCodeFrameError(
+      'ZACS StyleSheets can only contain simple style properties with properties like so: `{ backgroundColor: \'red\', height: 100 }`. Other syntaxes, like `[prop]:`, `"prop": `, `...styles` are not allowed.',
+    )
+  }
+
+  // TODO: Validate values
+}
+
+// TODO: Find a JS-to-CSS package
+// Inspiration: https://github.com/giuseppeg/style-sheet/blob/master/src/data.js
+const capitalRegex = /([A-Z])/g
+const cssCaseReplacer = (match, letter) => `-${letter.toLowerCase()}`
+function encodeCSSProperty(property) {
+  return property.name.replace(capitalRegex, cssCaseReplacer)
+}
+
+function encodeCSSStyle(property) {
+  return `  ${encodeCSSProperty(property.key)}: ${property.value.value};`
+}
+
+function encodeCSSStyleset(styleset) {
+  const { name } = styleset.key
+  const attributes = styleset.value.properties.map(encodeCSSStyle).join('\n')
+  return `.${name} {\n${attributes}\n}`
+}
+
+function encodeCSSStyleSheet(stylesheet) {
+  const stylesets = stylesheet.properties.map(encodeCSSStyleset).join('\n\n')
+  return `${stylesets}\n`
+}
+
+function transformStyleSheet(t, state, path) {
+  const { node } = path
+  const { init } = node
+  const { arguments: args } = init
+
+  validateStyleSheet(t, args, path)
+
+  const stylesheet = args[0]
+  const css = encodeCSSStyleSheet(stylesheet)
+
+  const preparedCss = `\n${css}ZACS_MAGIC_CSS_STYLESHEET_MARKER_END`
+  const magicCssExpression = t.expressionStatement(
+    t.taggedTemplateExpression(
```

Comment:
Any reasons for using a template literal instead injecting the CSS between two comments? Eg.:

```
// ZACS_MAGIC_CSS_STYLESHEET_MARKER_START
...
// ZACS_MAGIC_CSS_STYLESHEET_MARKER_END
```

Reply:
Comments are unreliable as far as AST transforms are concerned. Babel is known to mess up their formatting… Also, comments are not AST nodes, only annotations to preceding/following nodes. Also, a function call (or tagged template literal, same thing) will reliably crash the app if the webpack extraction to CSS fails. If I simply left a comment with the CSS to extract, or assigned it to a string constant, and webpack wasn't configured properly, then the failure would be silent or it would fail in some random place when `style.foo` didn't return a good value.